### ISSUE TITLE
[Feature Request] Function Granular Access

### ISSUE THREAD

**Issue by ly3xqhl8g9** [microsoft/TypeScript#49601]
# Suggestion

## 🔍 Search Terms

+ function granular access
+ function permissions
+ function sandbox

## ✅ Viability Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

## ⭐ Suggestion

Granular controls for functions to prevent/allow access to filesystem, network, and more.

A possible interface:

``` typescript
interface FunctionGranularAccess {
    network?: boolean | string | string[]; // allows network access for all (boolean), for one (string), for many (string[])
    filesystem?: boolean | string | string[]; // allows filesystem access for all (boolean), for one (string), for many (string[])
    exterior?: boolean; // allows access to functions/variables outside of the function's scope
    environment?: boolean; // allows access to process environment
    imports?: boolean; // allows the function to import dynamically
    dependencies?: string[] // specify the dependencies which can be used by the function
}
```

A possible syntax:

``` typescript
function name {granularAccess} (parameters) {body}
```

where `granularAccess` is a `TypeScript object` implementing the `FunctionGranularAccess` interface.

Due to the nature of `TypeScript`, this can be only a transpile-time feature: the resulting `JavaScript` is left unchanged, and only the developer is alerted of any granular transgressions.

## 📃 Motivating Example

Example of a simple `add`er function without any access:

``` typescript
/**
 * The function adds the `first` value with the `second` value.
 *
 * @param first
 * @param second
 * @returns
 */
function add {
    network: false,
    filesystem: false,
    exterior: false,
    environment: false,
    imports: false,
    dependencies: [],
} (
    first: number,
    second: number,
): number {
    return first + second;
}
```

[more examples](https://github.com/ly3xqhl8g9/papers/tree/master/fcd113%2C%20Function%20Granular%20Access)

## 💻 Use Cases

Extended programs and especially programs which are intended for extension and composition such as libraries require special care in terms of security.

The `NPM` ecosystem is only growing, and so is the dependency tree of even the simplest package. Having granular access over the permissions of each function, even if only at transpile-time, would give the developer a greater chance of defending against dependency changes/attacks and also enhances the design intent behind a function definition for non-adversarial, collaborative environments.

**Comment by fatcerberus** [comment]
This is an interesting idea, but probably *way* out of scope for TypeScript.

**Comment by MartinJohns** [comment]
TypeScript has no knowledge whether something is a network call or file access.

**Comment by fatcerberus** [comment]
@MartinJohns The way I read it is that functions would be annotated with that information and the compiler would verify.  So if you mark a function as `network: false`, you couldn't call ones marked `network: true` from inside it without getting an error.

**Comment by ly3xqhl8g9** [comment]
@fatcerberus Precisely, the following would throw an error:

``` typescript
import fetch from 'node-fetch';


function throwsError {
    network: false,
} () {
    fetch('https://github.com/');
}
```

The idea is nothing new: `Deno` has [file runtime permissions](https://deno.land/manual/getting_started/permissions), `Haskell` has [IO functions](https://wiki.haskell.org/Introduction_to_IO), and probably more languages have some support. The question is if `TypeScript` could support bringing the access controls to the function level and make the control really granular.

Yes, runtime permissions would always be better, since the forbidden functions could be hidden in some external object as below. Although even then the external scope access could be forbidden.`TypeScript` could do this perhaps easily by limiting the access of the function to the `Abstract Syntax Tree`'s external `ElementAccessExpression`s (image below or [link](https://ts-ast-viewer.com/#code/JYWwDg9gTgLgBAMwM6KhEcDkA7CATAUwC5lMBuAKCoGMJsl4CAPGAqbAQwBsB5AIwBWBavAC8cAN4U4MxEgA0FAL6UKCAK7YRwOnAbAuXACoALNAHckAUShooACgCUk6bOat23fkJEBtTKQAugB01FAEHKwAShF4AMow4Rwg9phsdpiOlEoUQA)).


``` typescript
import fs from 'node:fs';


function throwsError {
    filesystem: false,
} () {
    fs.createReadStream('error');
}


const externalObject = {
    fs,
};

function perhapsHardToCatch {
    filesystem: false,
} () {
    externalObject['fs'].createReadStream('error');
}


function stillThrowsError {
    exterior: false,
} () {
    externalObject['fs'].createReadStream('error');
}
```

![AST](https://user-images.githubusercontent.com/13864427/174521577-f58a8d08-a27e-4eb3-900b-573e7f57a763.png)

**Comment by Josh-Cena** [comment]
How does this play with external dependencies? If I have something in node_modules with

```ts
declare function iPromiseImGood {filesystem: false} ();
```

How do you know in practice if the declaration is being honest?

**Comment by ly3xqhl8g9** [comment]
@Josh-Cena A naive implementation, as I see it now for the `filesystem` example, would merely check for calls to `node:fs` or similar. 

Ultimately, since `TypeScript` is a transpile-time checker, you would trust the declaration just as you trust any regular type. The same way as when you `import React from 'react'` and you trust that there really is a `React.createElement` function.

**Comment by Josh-Cena** [comment]
Your proposal is about supply chain/dependency safety. It is basically moot if you have to rely on trusting whoever is publishing the package to also publish "honest" declarations.

**Comment by ly3xqhl8g9** [comment]
@Josh-Cena However, keep in mind that the `Function Granular Access` controls can also be set up by you when wrapping the external function, consider this:

``` typescript
import {
    iPromiseImGood,
} from '@BadActor/badPackage';


function myCodebase {
   filesystem: false,
} () {
    iPromiseImGood();
}
```

In this case I would trigger the transpilation checks from my own codebase, regardless of the granular controls of the bad `iPromiseImGood` function. The transpiler would look at the fully expanded `AST` of the `iPromiseImGood` function and ensure the granular checks are respected to the best of its ability.

**Comment by Josh-Cena** [comment]
> The transpiler would look at the fully expanded AST of the iPromiseImGood function and ensure the granular checks are respected to the best of its ability.

The transpiler doesn't have access to the type information. You would need to take this proposal to ECMAScript instead if you want it to have actual compile-time semantics, because otherwise any transpiler would simply strip it away.

Remember that type-checking and transpilation are two steps, and the transpiler won't take any AST information into account unless encoded in ECMAScript semantics.

**Comment by ly3xqhl8g9** [comment]
@Josh-Cena By transpilation I meant the entire process when running `tsc index.ts`, and yes, it would happen in parallel with the type cheking. Once `TypeScript` exposes the granular controls in the declaration of each function, the bundler can take them into account, sure.

**Comment by Josh-Cena** [comment]
> By transpilation I meant the entire process when running `tsc index.ts`

`tsc` never runs on node_modules—unless you actually publish your TypeScript source instead of compiled JavaScript output, and ask all users to compile it themselves, which I don't think any sane package is doing. Instead, `tsc` only reads the _declarations_, which is why I'm asking how well it plays with declarations.

**Comment by ly3xqhl8g9** [comment]
@Josh-Cena Yes, I also don't know. `Feature Request` is too strong in the title, perhaps `Somewhat of a Proposal` would have been more adequate.

Maybe a comments protocol with a runtime validator is the easier/better way, something along the lines of:

``` typescript
/*
 * access: 
 *   network: false
 */
function adder(
    a: number,
    b: number,
): number {
   return a + b;
}
```

and then after `tsc index.ts` you would run:

```
validate index.ts && node index.js
```

**Comment by MartinJohns** [comment]
> Precisely, the following would throw an error:

This would first require all packages like `node-fetch`, `fs`, etc. to add such annotations. I can't see this realistically happen.

**Comment by ly3xqhl8g9** [comment]
@MartinJohns Not at all, it should be enough to restrict the function only within my codebase and the validator would check all the calls made from within that function respect the granular restrictions. The harder part seems to me to statically classify if a function call is a network call, a filesystem call, etc.

**Comment by MartinJohns** [comment]
> The harder part seems to me to statically classify if a function call is a network call, a filesystem call, etc.

That's exactly what I meant. The libraries need to provide this information.

**Comment by ly3xqhl8g9** [comment]
@MartinJohns However, as pointed out about by Josh-Cena you cannot trust the libraries. The validation must be performed each time before running the code or at least after each dependencies update, doing a full-depth call-stack analysis.

**Comment by MartinJohns** [comment]
So... the entire proposal is moot. Libraries only provide declaration files, the compiler doesn't see the actual source code behind it - there's nothing to run a "full-depth call-stack analysis" on. The library, e.g. `node-fetch`, must provide the information that `fetch()` is a network call, because TypeScript itself has no knowledge about these things.

It's very unclear how you expect this to work at all.

**Comment by ly3xqhl8g9** [comment]
@MartinJohns Pretty much in the same way [Deno](https://deno.land/manual/getting_started/permissions) does it, just at the function-level, not at file-level. For the network example, `TypeScript` could check that no function from `net`, `http`, `https`, etc. is called if granular access `network: false`. For filesystem, no function call from `fs`, and so on. You run the call-stack analysis on the source itself, yours + dependencies. 

Nevertheless, as mentioned above, maybe a third-party utility checker/validator with restrictions specified as comments is the better way to go, even if having syntax sugar in the language is a better developer experience.

**Comment by Josh-Cena** [comment]
Deno is a runtime, and callstack analysis is also runtime. It seems you are proposing a runtime feature instead of a compile-time one?

**Comment by ly3xqhl8g9** [comment]
@Josh-Cena Yes, the proposal is to do the analysis statically at compile/type-check/transpile-time.

@MartinJohns The `TypeScript` [compiler/parser](https://github.com/microsoft/TypeScript/blob/96d298c4bcfc5b3ff2333ce066f5e0a5c80fc844/src/compiler/parser.ts#L108) does see the code from what I understand and the `SyntaxKind` would require extension in order to classify `NetworkFunctionDeclaration`, `FileSystemFunctionDeclaration`, etc. No module name hardcoding required.

I am not implying in the proposal that `TypeScript` is already set up for this kind of analysis, hence why it is merely a proposal.

**Comment by MartinJohns** [comment]
> The `TypeScript` [compiler/parser](https://github.com/microsoft/TypeScript/blob/96d298c4bcfc5b3ff2333ce066f5e0a5c80fc844/src/compiler/parser.ts#L108) does see the code from what I understand

You misunderstand. We're talking about **libraries**. Libraries provide [declaration files](https://www.typescriptlang.org/docs/handbook/declaration-files/introduction.html), not sources, which provide the information "I have a method foo", but it does provide the information what the method "foo" actually does.

This is the information we have in declaration files:

```typescript
declare function foo(): void;
```

But for your proposal to work you'd need to know the implementation of the method, which is not available for libraries. The method may call `fs`, it may not. The compiler doesn't know.

**Comment by ly3xqhl8g9** [comment]
Yes. But what is stopping the compiler from following that function declaration `foo` to the source file `foo.js` and check the calls `function foo()` makes? It's not implemented sure, but no fundamental obstacle as far as I can see. Hence what the proposal asks: is there such a fundamental obstacle?

**Comment by Josh-Cena** [comment]
Yes, because not many projects publish with declaration maps—even TypeScript's own `go to source definition` is only a "guess".

**Comment by ly3xqhl8g9** [comment]
Sure, we always deal with imperfect information. Security is never about certainty, but mitigated risk.

Yes, hardcode is clearly a no-way. I was thinking more along the lines of checking for system calls somehow à la `vm.runInNewContext`.

Nevertheless, thank you for the discussion. The idea remains, perhaps for another language. I will close now.
